pkg/scheduler/algorithm/priorities/resource_bin_packing.go

```diff
@@ -0,0 +1,240 @@
+	allocatable := nodeInfo.AllocatableResource()
+	used := nodeInfo.RequestedResource()
+	var requested schedulercache.Resource
+	var requestedInit schedulercache.Resource
+	switch resource {
+	case v1.ResourceCPU:
+		{
+			for i := range pod.Spec.Containers {
+				container := &pod.Spec.Containers[i]
+				cpu, _ := priorityutil.GetNonzeroRequests(&container.Resources.Requests)
+				requested.MilliCPU += cpu
+			}
+			for i := range pod.Spec.InitContainers {
+				container := &pod.Spec.InitContainers[i]
+				cpu, _ := priorityutil.GetNonzeroRequests(&container.Resources.Requests)
+				if requestedInit.MilliCPU < cpu {
+					requestedInit.MilliCPU = cpu
+				}
+			}
+			if requestedInit.MilliCPU > requested.MilliCPU {
+				requested.MilliCPU = requestedInit.MilliCPU
+			}
+
+			//println("CPU used: %d, requested: %d, available: %d",
+			//	nodeInfo.NonZeroRequest().MilliCPU, requested.MilliCPU, allocatable.MilliCPU)
+			return float64((nodeInfo.NonZeroRequest().MilliCPU+requested.MilliCPU)*schedulerapi.MaxPriority) / float64(allocatable.MilliCPU)
+
+		}
+
+	case v1.ResourceMemory:
+		{
+			for i := range pod.Spec.Containers {
+				container := &pod.Spec.Containers[i]
+				_, memory := priorityutil.GetNonzeroRequests(&container.Resources.Requests)
+				requested.Memory += memory
+			}
+			for i := range pod.Spec.InitContainers {
+				container := &pod.Spec.InitContainers[i]
+				_, memory := priorityutil.GetNonzeroRequests(&container.Resources.Requests)
+				if requestedInit.Memory < memory {
+					requestedInit.Memory = memory
+				}
+			}
+			if requestedInit.Memory > requested.Memory {
+				requested.Memory = requestedInit.Memory
+			}
+			//println("Memory used: %d, requested: %d, available: %d",
+			//	nodeInfo.NonZeroRequest().Memory, requested.Memory, allocatable.Memory)
+			return float64((nodeInfo.NonZeroRequest().Memory+requested.Memory)*schedulerapi.MaxPriority) / float64(allocatable.Memory)
+		}
+
+	case v1.ResourceEphemeralStorage:
+		{
+			for i := range pod.Spec.Containers {
+				container := &pod.Spec.Containers[i]
+				if quantity, ok := container.Resources.Requests[v1.ResourceEphemeralStorage]; ok {
+					requested.EphemeralStorage += quantity.Value()
+				}
+
+			}
+			for i := range pod.Spec.InitContainers {
+				container := &pod.Spec.InitContainers[i]
+				if quantity, ok := container.Resources.Requests[v1.ResourceEphemeralStorage]; ok {
+					if requestedInit.EphemeralStorage < quantity.Value() {
+						requestedInit.EphemeralStorage = quantity.Value()
+					}
+				}
+			}
+			if requestedInit.EphemeralStorage > requested.EphemeralStorage {
+				requested.EphemeralStorage = requestedInit.EphemeralStorage
+			}
+			return float64((used.EphemeralStorage+requested.EphemeralStorage)*schedulerapi.MaxPriority) / float64(allocatable.EphemeralStorage)
+
+		}
+	default:
+		if v1helper.IsScalarResourceName(resource) {
+			if podRequestsResource(*pod, resource) {
```

Comment:
prefer to cache that in a map; so we do need to go through all resources of container every time. And I'd prefer to add those logic to common part.

Reply:
The issue with traversing all the resources once is that it just makes the function very complicated as every resource score calculation is different and also the way we handle init containers is different. Right now resource calculation takes n*m where n is the number of resource and m is the number of containers which will still be the same as we would need to have another for loop inside every container for each resource type. I just feel this makes the code more readable.